### DISCUSSION TITLE
Add color and icon to the CalendarEntity

### DISCUSSION THREAD

**Discussion by Misiu** [home-assistant/architecture#883] [Entity Models] [answered]
## The goal
be able to specify color and icon per calendar.

## Current situation
we are unable to specify color and icon per calendar

Currently, we are unable to set color and icon per calendar.
It has been requested multiple times:
https://community.home-assistant.io/t/google-calendar-color/278352
https://community.home-assistant.io/t/calendar-add-on-improvements-color-and-calendar-selection/523407
https://github.com/home-assistant/frontend/issues/5975
https://github.com/home-assistant/frontend/discussions/11262
https://community.home-assistant.io/t/google-calendar-respect-color-individual-calendar-coding-in-google-calendar/523404
https://community.home-assistant.io/t/local-calendar-enhancements/510905
https://community.home-assistant.io/t/choose-custom-colors-for-calendar-google/391003
https://community.home-assistant.io/t/google-calendars-settings-colors-for-tags/435449
https://community.home-assistant.io/t/adjusting-colors-google-calendar-integration/461751

## Proposal
Add color property to [CalendarEntity](https://github.com/home-assistant/core/blob/dev/homeassistant/components/calendar/__init__.py#L343) and expose Color and Icon in Config Flow for local calendars (I didn't check other calendar types and APIs, so I'm not sure if, for example, Google returns calendar colors)
CalendarEntity extends the Entity class which has the Icon property, but the color should be a CalendarEntity property.

**Comment by frenck** [comment]
colors are not part of entities in Home Assistant in general. Icons, however, are. Calendar entities can already have custom icons and provide icons. Nothing is needed for that. Could you explain what I am missing?

Additionally, your question about exposing icons in the config flow, is not part of the architectural discussion (it is an integration-specific thing) and also not part of an integration option (thus should not have to be in the config flow).

I'm not sure if making an exception for adding colors in a single type of entity makes sense tbh.

../Frenck

**Comment by Misiu** [comment]
The idea is quite simple:
I'd like to be able to specify color per calendar. I searched and found that others have the same problem - this currently can't be done.
I'm migrating from https://github.com/bruxy70/Garbage-Collection to the local calendar and thought it would be good to be able to specify the colors of calendars the same as are on trash bins.
For some reason Google added color pickers to their calendar:
![image](https://user-images.githubusercontent.com/1741838/225140783-504d75c3-b3b4-4056-b241-761001c109ac.png)

> Calendar entities can already have custom icons and provide icons. Nothing is needed for that. Could you explain what I am missing?

Ok, I found this, we can set icons for entities (calendar is an entity), but the calendar view isn't showing them. This is a frontend issue, I'll address it there.

> colors are not part of entities in Home Assistant in general.

I agree with that, but if this won't be a property of CalendarEntity then this won't work right? The color must be persisted not only in the browser)

**Reply by Misiu** [reply]
@allenporter as you created the local calendar I'd like to ask for your opinion.
Google has API to get colors (https://developers.google.com/calendar/api/v3/reference/calendarList/list?hl=en), so if we add support for colors, then not only local calendar can benefit.

**Reply by frenck** [reply]
Please withhold from pulling people into conversations. You might have the right intentions, but is may come across as demanding their attention. Thanks 👍

**Reply by allenporter** [reply]
FWIW I've been reading this discussion in my inbox and didn't have anything to add beyond what was said already by @frenck 

(Similarly i've seen https://github.com/home-assistant/core/pull/88339 and am not feeling super excited about it in the current form)

**Comment by frenck** [comment]
So, got a question on Discord related to my comment too.

Let me start with: I'm not against it, but I'm not sure if it makes sense to put colors on calendar entities only.

There are some concerns, in general, to take into account. For example, themes and theme colors. The chosen background color may be problematic with the colors provided, the `light` entity already has these issues (and works around it partially).

On the other end of the spectrum, it might make sense to apply a "color" (or colors) to all entities. The reasoning for that originates from the general design overhaul taking place right now. Take, for example, the new tile card, having colors on the entity level might be a good fit for that.

But it is not just the tile card, entity colors are in many more places nowadays.

Looking at the recently introduced state colors for entities, a color setting like the suggested might conflict with that. What would be the difference in the color of the entity when the calendar entity is on or off? And if we provide colors on the entity level, what would that mean for entities having different state colors for their different states?

Hope this makes my train of thought clearer, why I think adding color to a calendar entity might sound great from the "Calendar" perspective, but might need more tinkering/UX from a full Home Assistant perspective.

../Frenck

**Reply by allenporter** [reply]
This matches my view. I'd like to see a proposal here that considers the broader implications. In my own home assistant instance i see various colors in the UI for entities in graphs and things or various colors on cards and such and so having a one-off for calendar isn't immediately obvious to me as an answer here.

**Reply by Misiu** [reply]
That's why I started this discussion. I have a use case, but I don't have the knowledge to be able to say whether adding a color property to CalendarEntity or Entity is a good or a bad idea and what consequences it might have.
Adding color to Entity would open a lot of possibilities, but it also brings a lot of questions.

> Hope this makes my train of thought clearer, why I think adding color to a calendar entity might sound great from the "Calendar" perspective, but might need more tinkering/UX from a full Home Assistant perspective.

Sadly I can't say much about UX and the internals of Home Assistant, but at least we are talking about possible change.

**Reply by jonkristian** [reply]
I kinda disagree with the basis of those arguments about possibly themes not working with the colors in the calendar or that it might not look good, because there are several ways to account for that when styling, and wouldn't it be up to the theme authors to make the appropriate styling to account for these changes? State on/off and how colors work is probably up to the theme, but one idea could be using rgba to determin brightness... or these days you could use css filters.

Also I just noticed this, but it seems the official calendar card doesnt even work with the colors from the calendar because they are determined just by index and by selectively adding calendar entities (which you can) those indexes will shift and then provide the wrong color.

**Reply by arrisar** [reply]
While I agree that a one-off here might not be the _long-term_ answer, there is an obvious use case for selecting the colour of your calendar.

Progressive enhancement, I believe, is the best path for features like this. If we can scope this solution to the  `CalendarEntity` to begin with, then it provides a great reference point for further discovery and broader application.

---

My apologies to necro this discussion, but as far as I can tell it's the most poignant place. Please feel free to redirect me if there's a better one.

The [PR that Allen referenced](https://github.com/home-assistant/core/pull/88339) is indeed a bit naïve in implementation, but would it be appropriate to submit another?

**Reply by allenporter** [reply]
My suggestion would be to work through the discussion points raised and make a proposal to address them directly or reason through why they're OK,. I don't think that sidestepping the points raised is the way to go in the name of iteration. My impression is that custom cards and custom panels can be used for iteration so there is a solution for iteration.

**Reply by Jonny-T-cyber** [reply]
> My suggestion would be to work through the discussion points raised and make a proposal to address them directly or reason through why they're OK,. I don't think that sidestepping the points raised is the way to go in the name of iteration. My impression is that custom cards and custom panels can be used for iteration so there is a solution for iteration.

I completely agree.
I hope for a solution that includes personalized cards in which you can choose the colors and icons of the calendar

**Comment by robloh** [comment]
I tried to thing of a more broader use and the only thing I could think of was color for entities could be really helpful with graphs as well (for all entities). But perhaps that's better solved on the graph cards.

On the other hand I do find colors on my iOS calendars extremely helpful with 6 calendars to quickly see at a glance what's important and what's not, just by color. So I think there's definitely a use case for calendars.

Perhaps this is a silly thing to mention, but of course certain light entities can have color as well ;)

**Comment by sym0nd0** [comment]
Just to add to the convo, I would love to see something like this (not mine) from the forum https://community.home-assistant.io/t/local-calendar-enhancements/510905.

Like many, I'm just now moving over to Local Calendar following the Garbage Collection customer integration being deprecated, due to Local Calendar having most of (if not all of) the same functionality. No problem there for me, Local Calendar works really well. I'd just like to be able to choose the colour associated for each calendar myself (mostly as I have four bins and would like to match the calendar colour to the bin...yeah, daft I know 😂).

I think the addition of Material Design icons, maybe utilising the same icon picker like with entities, and user definable colour (colour wheel, hex or the like) would fit with the way HA UI has been progressing over the last couple of years.

All that said, I am totally ignorant to just how difficult to achieve this _simple_ idea would be.

**Comment by Brtrnd** [comment]
Hi
Calendars already have colours, but you can't pick the color you like. I believe allowing users to choose the colours is not a large change.
Please see my reply in this topic: https://community.home-assistant.io/t/google-calendar-color/278352/11 where I propose to change line 148 of of [this file](https://github.com/home-assistant/frontend/blob/e1be4751a1cc3985611ac93330ef7ebb2d92752e/src/data/calendar.ts#L4) to include an if statement
.map((eid, idx) => ({
      entity_id: eid,
      name: computeStateName(hass.states[eid]),
      backgroundColor: hass.states[eid].colour if (hass.states[eid].colour)  else getColorByIndex(idx) ,
    }));



I don't think this would break stuff that would already be broken with the default colors, but it would add consistancy between my google calendar and homeassistant.
But I'm not a coder and I'm not an expert.

**Comment by Paktas** [comment]
+1 to being able to pick calendar icon color

**Comment by andy-vdg** [comment]
+1 following

**Comment by averater** [comment]
Is there any progress? I think it sounds Great to have colors for all entities. I have been missing that. Especially for lights where the blue icon and a blue light on looks similar.

If there is a separate pr for adding colors then please linkt it.

**Comment by blubkens** [comment]
I ran into the same issue as misiu. 
The wife uses pink, myself blue and shared is green. 
In HA my wife is blue, i’m pink and shared is light blue. 
Colors are there but assigning them is not possible. 
Would love to have the calendar on the dashboard but at the moment it’s not usable for us.

**Comment by jprocket45** [comment]
well, I half to agree with colors in calendar sadly I would love to use HA for everything but my wife disagrees if she can't color code the calendar where not using any calendar inside HA seems a lot of people have issues with this... So I guess I am stuck using actual google calendar.

**Reply by doctorkb** [reply]
It isn't completely native, but https://github.com/FamousWolf/week-planner-card lets you specify colors for each calendar.

**Comment by KGS501** [comment]
One problem I am seeing in this conversation is, that it seems to be a big change to add some sort of color to the entity itself like a native property. Maybe it would be an idea to use some sort of attribute of the calendar entity to assign color. Attributes are in my experience quite versatile and not standardized, so a simple `color: rgb(255,0,0)` or something like this as an attribute, which can be edited by the user through some sort of frontend ui might be a more feasible way of implementing this, as I personally think too, that this implementation would be very helpful to distinguish different calendar events from multiple calendars at a quick (and more importantly for the specific user more intuitive) glance

**Comment by doctorkb** [comment]
The biggest issue with *not* doing it is that it results in inconsistent behaviour.  If I create three Calendar cards, the first two with one calendar entity each, and the third with both of those calendar entities, the first two cards will have blue items for their entity.  The third card will have blue and purple items.  

For a UX that matches what people have become used to with other productivity tools (gCal, Outlook, etc.), that second entity should *always* be purple (in this particular example) and not change colours based on the order in which it happened to be listed in a display model.  This issue becomes much worse when we're not talking about two calendars over three cards, but a dozen or so (personal and work for a few family members, plus others for extracurricular activities).

**Comment by CMDR-Sloma** [comment]
>  So, got a question on Discord related to my comment too.
> 
> Let me start with: I'm not against it, but I'm not sure if it makes sense to put colors on calendar entities only.
> 
>   There are some concerns, in general, to take into account. For example, themes and theme colors. The chosen background color may be problematic with the colors provided, the light entity already has these issues (and works around it partially).
> 
>   On the other end of the spectrum, it might make sense to apply a "color" (or colors) to all entities. The reasoning for that originates from the general design overhaul taking place right now. Take, for example, the new tile card, having colors on the entity level might be a good fit for that.
> 
>   But it is not just the tile card, entity colors are in many more places nowadays.
> 
>   Looking at the recently introduced state colors for entities, a color setting like the suggested might conflict with that. What would be the difference in the color of the entity when the calendar entity is on or off? And if we provide colors on the entity level, what would that mean for entities having different state colors for their different states?
> 
>   Hope this makes my train of thought clearer, why I think adding color to a calendar entity might sound great from the "Calendar" perspective, but might need more tinkering/UX from a full Home Assistant perspective.
> 
>    ../Frenck

Correct me if I am wrong here, but I find that your perspective comes from the coder point of view and the explanation is really technical, which really does not interest me as a user. I don't want to come as rude or anything, so in other words, I'd just love the option to customize colours of the calendars or colours within home assistant easily. From a technical point of view, I will take any solution that will work for you as long as the solution will allow me to achieve what I want. It's just at the moment the way it works is terrible, sometimes you end up white text on yellow background because HA decided to assign these two colour pairs arbitrarily to a new calendar I created.

I think at the moment colours in themes / calendars are completely "broken", there is no easy way to change them except for tinkering with yaml (themes) which, from my perspective, is not an easy task.

I hope this makes sense.

**Comment by smplcrtrs** [comment]
I use a Google Calendar as my primary HA calendar. On the Google side of things, I am able to assign a colour to each member of the family, so we don't have to use an individual calendar for each of us. This vital bit of information is lost when an event or calendar is viewed in HA. I'd really like the ability to add `colorId` as one of the attributes piped over. Using it could be optional, but I'd find it so useful to colour code my calendar in HA. Translating colours added to events created in HA for importing into the Google Calendar might be more difficult, but it's not insurmountable.

https://developers.google.com/workspace/calendar/api/v3/reference/events

https://developers.google.com/workspace/calendar/api/v3/reference/colors

**Comment by frenck** [comment, marked as answer]
We have discussed this topic and proposal in an architectural meeting today. As it currently stands, the customization of the icons is already possible, as calendars are entities.

As for the color, we are okay with adding this as a property to the CalendarEntity and to the Home Assistant frontend so that an integration can provide it and the user can customize it.

../Frenck

**Reply by smplcrtrs** [reply]
Frenck, you are a legend. Thank you!

**Reply by blubkens** [reply]
Great work team!

**Reply by allenporter** [reply]
I was doing more looking today and determined that `color` is part of rfc5545
https://datatracker.ietf.org/doc/html/rfc7986#section-4 for both the calendar.  (For future reference, also it's part of individual events or todos)

One quick clarification as we go into [implementation](https://github.com/home-assistant/core/pull/145606), this is what I have in mind:
- For reading this, it makes sense to set as an attribute on the calendar. This can be supplied by the integration (e.g. Google calendar returns color attribute)
- For setting it, it can be a new method e.g. `asyc_set_color` and  the feature `CalendarEntityFeature.SET_COLOR_ON_CALENDAR`.  This is verbose intentionally to clarify setting on the calendar vs item, similar to the Todo list entity feature verbosity).  This implies that the color in the calendar UI with a new bespoke API.
- An alternative is that it can be part of the options flow. This is the current approach in the PR for local calendar, but I am realizing maybe it should settable on the entity itself instead.  The options flow is nice to avoid introducing a new concept, but i'm thinking given the likely desired UX and RFC above it makes sense on the entity itself.

Could you provide some additional guidance based on the core group decision/discussion if you had something specific in mind here?

**Reply by smplcrtrs** [reply]
Hi Allen, Thank you for your thoughts here. I feel (bearing in mind I'm more an end user with some knowledge of code rather than someone with the in-depth skills of you and the rest of the team) that setting it as an attribute would be the best solution.

I might be wrong, but I'd think that that would allow integrations for various different calendar back-ends (Google, 365, iCal etc) to communicate colours in both directions with HA. Hopefully all the major 3rd parties respect RFC standards.

I really appreciate the work you've all put in with this, and I'm excited to see how it goes! 

Let me know if you need a useful idiot to test things out :)

**Reply by smplcrtrs** [reply]
PS: Do you see this attribute eventually evolving into other entities beyond calendars? And does there need to be consideration of accessibility and colour-blindness?

**Reply by Misiu** [reply]
@allenporter, how can we move this forward? IMHO, config flow and options flow work, so we have a base implementation. But it would be great to have color on the item level. For example, right now I have 4 calendars for trash day (each type has its own calendar), with my change, I will finally be able to fix the color per calendar, but having the option to set color per entry would allow me to have a single calendar or even show important events using different colors.
Right now, I'm using the awesome [Calendar Card Pro](https://github.com/alexpfau/calendar-card-pro), which fixes all the issues, but having more control over the colors in the core would be great.

**Reply by allenporter** [reply]
> @allenporter, how can we move this forward? IMHO, config flow and options flow work, so we have a base implementation. 

I would recommend proposing something that addresses what I'm talking about?  My suggestion would be to go with calendar level attribute, and building on past decisions we've made here and work through the frontend details to make it happen.

@emontnemery if you have any input from the core group that would be helpful? since the approval didn't get specific.  Otherwise we can go with my recommendation if I don't hear back.

**Reply by emontnemery** [reply]
@allenporter we're suggesting:
- Store the color as an entity option in the entity registry, under the calendar domain. Exposing it as a state attribute does not seem very useful
- Let integrations read the color from the calendar service (or from the user for local calendars) when the calendar entity is first created, but after that we don't expect sync between the service and display in Home Assistant
- Changing the color is done by writing to the entity registry, not by a service

**Reply by Misiu** [reply]
Core PR: https://github.com/home-assistant/core/pull/145606
Frontend: https://github.com/home-assistant/frontend/pull/28882

**Reply by Misiu** [reply]
@emontnemery I've also added changes to Google Calendar integration: https://github.com/home-assistant/core/commit/b085d3701159d4544ee0a830ab3b9d507018bbc1. The only thing I needed to add is get_initial_entity_options.
Will create another PR after I discuss it more with Allen.

**Reply by Misiu** [reply]
Google Calendar support: https://github.com/home-assistant/core/pull/161671

**Comment by markusg1234** [comment]


**Reply by Soukyuu** [reply]
It's not like people work full time and get paid for this.

**Reply by tetele** [reply]
Anything keeping you from making improvements occur with your own hands? What's that? Lack of know-how? Lack of time? Day-to-day job? Other priorities? Sad, very sad...

**Reply by tetele** [reply]
Some get paid from subscriptions, correct. How long have you had your (voluntary) subscription for?

You are aware that architecture discussions are not feature requests, right? You do know that architecture discussions are contributing devs asking whether their contribution to alter the core is ok, right?

**Reply by parautenbach** [reply]
@markusg1234 let's try to stick to comments that can move this forward or that's actively contributing to the project. I don't think entitlement belongs here: Use the social category on the forum if you feel like ranting. Also have a look at the hundreds of changes being made every month to the project. I can understand that you care about a feature like this, but generalising in the way you are isn't helpful. 

Just because someone gets paid doesn't entitle you to demands. The core team works according to a roadmap and the rest of the changes are made by other maintainers and contributors – that work (read: do this for fun) for free.

**Comment by smplcrtrs** [comment]
I'm happy to wait as long as these things take. Development of HA is a hugely complex process, and there are higher priority things on the road map at the moment.
